rec: Implement EDNS0 padding (rfc7830) for outgoing responses

rgacogne opened the pull request:
### Short description
<!-- Write a small description of what this Pull Request fixes or provides, including the issue #s -->
The initial idea is to be able to add EDNS padding when answering to queries received over DoT or DoH. This PR allows adding this option based on the source (so you could use a different source when forwarding a query from dnsdist to the recursor, based on whether it was received over DoH or DoT), and optionally on whether the initial query had padding.

One of the challenge is that this duplicates the content of the packet cache to avoid mixing padded and non-padded responses.

### Checklist
<!-- please indicate if any of these things are done/included with this Pull Request. Not all boxes need to be checked for the Pull Request to be accepted -->
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [x] included documentation (including possible behaviour changes)
- [x] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

Habbie commented:
> One of the challenge is that this duplicates the content of the packet cache to avoid mixing padded and non-padded responses.

Without having read the code - the one thing our packet cache responses currently special-case is the QID, and the exact caPiTAlISation of the QNAME. Perhaps we can do the padding the same way?

Habbie commented on pdns/pdns_recursor.cc:
```diff
@@ -1172,9 +1179,24 @@ static void startDoResolve(void *p)
+          /* we should only pad on 'secure' cases to limit amplification:
+             - over TCP ;
+             - from 'known-good' sources (dnsdist for example).
+             We should also support a strict mode (only pad if the query has a padding option)
```
```suggestion
             TODO: We should also support a strict mode (only pad if the query has a padding option)
```

rgacogne replied:
It's actually already implemented, I'll update the comment :)

Habbie commented on pdns/pdns_recursor.cc:
```diff
@@ -1172,9 +1179,24 @@ static void startDoResolve(void *p)
+             - over TCP ;
+             - from 'known-good' sources (dnsdist for example).
+             We should also support a strict mode (only pad if the query has a padding option)
+             and a relaxed mode (pad if the client supports EDNS).
```
This reminds me of https://community.cloudflare.com/t/doh-service-adds-edns-padding-even-if-client-does-not-indicate-edns-support/84199 :)

rgacogne commented:
> (The next question then would be if what we put in the packet cache is padded or not; for 'padded', we could send a 'truncated' version when we're not padding; this might be cheaper than a resize when doing a 'padded' from a cached non-padded response. Or perhaps none of this matters because the QID/QNAME fixing already involves a copy, and we can fix that copy to already allocate the right padded/nonpadded size in one go.)

Fixing the padding is not that easy, because we also need to update the size of the OPT rdata length.

Habbie commented:
> Fixing the padding is not that easy, because we also need to update the size of the OPT rdata length.

Ah, yes :(

omoerbeek commented:
@rgacogne this PR is still in "Draft" status. It is already worth the time doing a review? Or is it better to wait?

rgacogne commented:
> @rgacogne this PR is still in "Draft" status. It is already worth the time doing a review? Or is it better to wait?

IMHO we need to decide whether we are interested in implementing this, and if we are what to do for the packet cache. It wouldn't be _that_ hard to add the padding on a packet cache hit, if we decide it's worth it.

rgacogne commented:
I'm wondering whether the packet duplication is a real issue. Until #8993 is merged we already store  different responses depending on the incoming padding anyway, and once it's merged we will only store two.

pieterlexis commented:
I've closed #10018 in favor of this one. @omoerbeek can you milestone this PR?

omoerbeek commented:
@rgacogne As there's been some interest in getting this, would you be able to get this into shape for 4.5? If not, we'll move the milestone, but if it is possible that would be nice.

rgacogne commented:
> As there's been some interest in getting this, would you be able to get this into shape for 4.5? If not, we'll move the milestone, but if it is possible that would be nice.

Sure, but the reason is stayed a draft PR for so long is that I would like some feedback on the general logic, and especially on what we should with regard to the packet cache.

omoerbeek commented on pdns/ednspadding.cc:
```diff
@@ -0,0 +1,31 @@
+std::string makeEDNSPaddingOptString(size_t bytes)
+{
+  std::string ret;
+  ret.resize(bytes, '0');
```
De we want zero or NUL?

rgacogne replied:
A `CharT` whose value value is `0`.

omoerbeek replied:
Form the RFC: `The PADDING octets SHOULD be set to 0x00.`.
So I think should be `0` (or `'\0'`and not `'0'`. The latter is ASCII zero with value `0x30`.

rgacogne replied:
Ouch, sorry, will fix!

omoerbeek commented on pdns/recursordist/docs/settings.rst:
```diff
@@ -548,6 +548,40 @@ found, the recursor fallbacks to sending 127.0.0.1.
+
+Whether to add EDNS padding to all responses (``always``) or only to the ones for padded queries (``padded-queries-only``, the default) coming from `edns-padding-from`_ sources.
+
+.. _setting-edns-padding-mode:
```
```suggestion
.. _setting-edns-padding-tag:
```

omoerbeek commented on pdns/recursordist/docs/settings.rst:
```diff
@@ -548,6 +548,40 @@ found, the recursor fallbacks to sending 127.0.0.1.
+-  One of ``always``, ``padded-queries-only``, String
+-  Default: ``padded-queries-only``
+
+Whether to add EDNS padding to all responses (``always``) or only to the ones for padded queries (``padded-queries-only``, the default) coming from `edns-padding-from`_ sources.
```
From this description it is not immediate clear if queries outside `edns-padding-from` are padded or not.

rgacogne replied:
This PR implements padding for outgoing responses, not for outgoing queries, perhaps this description should be clearer about that?

omoerbeek replied:
My comment should have read:
From this description it is not immediate clear if responses to queries outside edns-padding-from are padded or not.

rgacogne replied:
I have updated the description, please let me know if it's still unclear. It's sometimes hard to take the user point-of-view when you wrote you spent some time writing the actual code! :)

omoerbeek replied:
Yes, much more clear.

omoerbeek commented on pdns/pdns_recursor.cc:
```diff
@@ -1640,6 +1661,30 @@ static void startDoResolve(void *p)
+           RFC 7830) and SHOULD pad the corresponding response to a
+           multiple of 468 octets (see below).
+        */
+        static const size_t blockSize = 468;
```
Little use making a `const int` static here

omoerbeek reviewed (commented):
- I think the extra pressure on the packetcache is acceptable, especially since we now have #8993 indeed.
- If it does becomes a problem we can implement padding after packet cache retrieval. Wonder if it's worth the complexity though.
- Please make sure in the description of the PR that we're talking client side only

pieterlexis commented on pdns/recursordist/docs/settings.rst:
```diff
@@ -548,6 +548,40 @@ found, the recursor fallbacks to sending 127.0.0.1.
+
+``edns-padding-mode``
+---------------------
+.. versionadded:: 4.4.0
```
4.5.0

pieterlexis commented on pdns/recursordist/docs/settings.rst:
```diff
@@ -548,6 +548,40 @@ found, the recursor fallbacks to sending 127.0.0.1.
+-  Default: 7830
+
+The packetcache tag to use for padded responses, to prevent a client not allowed by the `edns-padding-from`_ list to be served a cached answer generated for an allowed one. This
+effectively divides the packet cache in two when `edns-padding-from`_ is used. Note that this will not override a tag set from one of the ``Lua`` hooks.
```
it might be good to expose whether or not a response would be padded in the `dq` object?

pieterlexis replied:
(if at all possible without making a trail of blood through the recursor)

rgacogne commented:
Rebased, fixing a few reported issues in the process. I'll see how hard it would be to make the padding status available to Lua hooks.

rgacogne commented:
Fixed the documentation as it turned out my memory was not right and the options provided by this PR are:
- Don't send padded responses (the default) ;
- Send padded responses to all queries from allow-listed (`edns-padding-from`) clients (`edns-padding-mode` set to `always`) ;
- Send padded responses to queries from allow-listed (`edns-padding-from`) clients provided that the query contains an EDNS Padding option (`edns-padding-mode` set to `padded-queries-only`).

Otto, does that seem OK to you?

I also added the ability to disable EDNS Padding from `gettag_ffi` and `DNSQuestion` based Lua hooks.

omoerbeek commented on pdns/pdns_recursor.cc:
```diff
@@ -5530,6 +5602,10 @@ int main(int argc, char **argv)
 
     ::arg().setSwitch("aggressive-nsec-cache-size", "The number of records to cache in the aggressive cache. If set to a value greater than 0, and DNSSEC validation is enabled, the recursor will cache NSEC and NSEC3 records to generate negative answers, as defined in rfc8198")="100000";
 
+    ::arg().set("edns-padding-from", "List of netmasks (proxy IP in case of XPF or proxy-protocol presence, client IP otherwise) for which EDNS padding will be enabled in responses, provided that 'edns-padding-mode' is set")="";
```
```suggestion
    ::arg().set("edns-padding-from", "List of netmasks (proxy IP in case of XPF or proxy-protocol presence, client IP otherwise) for which EDNS padding will be enabled in responses, provided that 'edns-padding-mode' applies")="";
```

omoerbeek commented on pdns/recursordist/docs/settings.rst:
```diff
@@ -610,6 +610,41 @@ found, the recursor fallbacks to sending 127.0.0.1.
+-  Comma separated list of netmasks
+-  Default: (none)
+
+List of netmasks (proxy IP in case of XPF or proxy-protocol presence, client IP otherwise) for which EDNS padding will be enabled in responses, provided that `edns-padding-mode`_ is set.
```
```suggestion
List of netmasks (proxy IP in case of XPF or proxy-protocol presence, client IP otherwise) for which EDNS padding will be enabled in responses, provided that `edns-padding-mode`_ applies.
```

omoerbeek reviewed (approved):
This time I compared the docs against the code and observed behaviour. It looks good with two small nits.
